 This notebook will be mainly used for the capstone project.

This project is focused  to predict the severity  a car accident based on the big data set(link below on "data" variable)

### Business Understanding:
As car accidents happenes on our daily lives, a ML Tool can help to understand how badly these "event" are. Taking into account different variables: road conditions, weather conditions, public light conditions, etc. May help to enhance the "awarnes" of the meaning "to drive". This document is directed to the comunity in general sense and can be used to enhance driving campaigns ads as reminder to drive carefully.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.utils import resample
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn import metrics

In [2]:
print('Hello Capstone Project Course!')

Hello Capstone Project Course!


In [3]:
data=pd.read_csv("https://s3.us.cloud-object-storage.appdomain.cloud/cf-courses-data/CognitiveClass/DP0701EN/version-2/Data-Collisions.csv")

/opt/conda/envs/Python36/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (33) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### Data understanding:
As you can notice, this dataset consists of 37 attributes; which can be splitted between independent variables and the 'SEVERITYCODE' as dependent variable(target). The objective is build a model to predict the 'SEVERITYCODE' based on those independent variables(note:we might exclude some of them due significance).
<ul>
    <li>To start some issues need to be fixed:
    <li>It has too many attributes (37)</li>
    <li>It is an unbalanced dataset (SEVERITYCODE)</li>
    <li>Some of the data is NaN</li>
</ul>

Considering the facts already shown, for the study would be selected as independent variables, Road Conditions, Liight Conditions, Weather and Speeding. These variables could provide major signifiance of probabilities of an accident may occur and how badly may resault(to predict SEVERITYCODE variable).

In [4]:
data.describe()

SEVERITYCODE              X              Y       OBJECTID  \
count  194673.000000  189339.000000  189339.000000  194673.000000   
mean        1.298901    -122.330518      47.619543  108479.364930   
std         0.457778       0.029976       0.056157   62649.722558   
min         1.000000    -122.419091      47.495573       1.000000   
25%         1.000000    -122.348673      47.575956   54267.000000   
50%         1.000000    -122.330224      47.615369  106912.000000   
75%         2.000000    -122.311937      47.663664  162272.000000   
max         2.000000    -122.238949      47.734142  219547.000000   

              INCKEY      COLDETKEY         INTKEY  SEVERITYCODE.1  \
count  194673.000000  194673.000000   65070.000000   194673.000000   
mean   141091.456350  141298.811381   37558.450576        1.298901   
std     86634.402737   86986.542110   51745.990273        0.457778   
min      1001.000000    1001.000000   23807.000000        1.000000   
25%     70383.000000   70383.000000   28667.000000        1.000000   
50%    123363.000000  123363.000000   29973.000000        1.000000   
75%    203319.000000  203459.000000   33973.000000        2.000000   
max    331454.000000  332954.000000  757580.000000        2.000000   

         PERSONCOUNT       PEDCOUNT    PEDCYLCOUNT       VEHCOUNT  \
count  194673.000000  194673.000000  194673.000000  194673.000000   
mean        2.444427       0.037139       0.028391       1.920780   
std         1.345929       0.198150       0.167413       0.631047   
min         0.000000       0.000000       0.000000       0.000000   
25%         2.000000       0.000000       0.000000       2.000000   
50%         2.000000       0.000000       0.000000       2.000000   
75%         3.000000       0.000000       0.000000       2.000000   
max        81.000000       6.000000       2.000000      12.000000   

        SDOT_COLCODE    SDOTCOLNUM     SEGLANEKEY  CROSSWALKKEY  
count  194673.000000  1.149360e+05  194673.000000  1.946730e+05  
mean       13.867768  7.972521e+06     269.401114  9.782452e+03  
std         6.868755  2.553533e+06    3315.776055  7.226926e+04  
min         0.000000  1.007024e+06       0.000000  0.000000e+00  
25%        11.000000  6.040015e+06       0.000000  0.000000e+00  
50%        13.000000  8.023022e+06       0.000000  0.000000e+00  
75%        14.000000  1.015501e+07       0.000000  0.000000e+00  
max        69.000000  1.307202e+07  525241.000000  5.239700e+06

In [5]:
data.head(5)

SEVERITYCODE           X          Y  OBJECTID  INCKEY  COLDETKEY REPORTNO  \
0             2 -122.323148  47.703140         1    1307       1307  3502005   
1             1 -122.347294  47.647172         2   52200      52200  2607959   
2             1 -122.334540  47.607871         3   26700      26700  1482393   
3             1 -122.334803  47.604803         4    1144       1144  3503937   
4             2 -122.306426  47.545739         5   17700      17700  1807429   

    STATUS      ADDRTYPE   INTKEY  ... ROADCOND                LIGHTCOND  \
0  Matched  Intersection  37475.0  ...      Wet                 Daylight   
1  Matched         Block      NaN  ...      Wet  Dark - Street Lights On   
2  Matched         Block      NaN  ...      Dry                 Daylight   
3  Matched         Block      NaN  ...      Dry                 Daylight   
4  Matched  Intersection  34387.0  ...      Wet                 Daylight   

  PEDROWNOTGRNT  SDOTCOLNUM SPEEDING ST_COLCODE  \
0           NaN         NaN      NaN         10   
1           NaN   6354039.0      NaN         11   
2           NaN   4323031.0      NaN         32   
3           NaN         NaN      NaN         23   
4           NaN   4028032.0      NaN         10   

                                          ST_COLDESC  SEGLANEKEY  \
0                                  Entering at angle           0   
1  From same direction - both going straight - bo...           0   
2                             One parked--one moving           0   
3                   From same direction - all others           0   
4                                  Entering at angle           0   

   CROSSWALKKEY  HITPARKEDCAR  
0             0             N  
1             0             N  
2             0             N  
3             0             N  
4             0             N  

[5 rows x 38 columns]

In [6]:
data['SEVERITYCODE'].value_counts()

1    136485
2     58188
Name: SEVERITYCODE, dtype: int64

The first step is to select the wanted attributes:

In [7]:
df=pd.DataFrame(data[['SEVERITYCODE','WEATHER','ROADCOND', 'LIGHTCOND','SPEEDING']])
df.head()

SEVERITYCODE   WEATHER ROADCOND                LIGHTCOND SPEEDING
0             2  Overcast      Wet                 Daylight      NaN
1             1   Raining      Wet  Dark - Street Lights On      NaN
2             1  Overcast      Dry                 Daylight      NaN
3             1     Clear      Dry                 Daylight      NaN
4             2   Raining      Wet                 Daylight      NaN

Next is to fill the Nan values, in order to place them through the study:

In [8]:
df["SPEEDING"].fillna("N", inplace = True)
df.head()

SEVERITYCODE   WEATHER ROADCOND                LIGHTCOND SPEEDING
0             2  Overcast      Wet                 Daylight        N
1             1   Raining      Wet  Dark - Street Lights On        N
2             1  Overcast      Dry                 Daylight        N
3             1     Clear      Dry                 Daylight        N
4             2   Raining      Wet                 Daylight        N

Then, the raw data has different characteristics, so in oorder to place a common scale, is necessary to transform it, this case would be through sequencial numbers like shown below

In [9]:
encoding_weather={'WEATHER':{'Clear': 1, 'Partly Cloudy':2, 'Overcast':3, 'Fog/Smog/Smoke':4, 'Severe Crosswind':5, 'Raining':6,'Snowing':7,'Sleet/Hail/Freezing Rain':8,'Blowing Sand/Dirt':9,'Unknown':10, 'Other':11}}
df.replace(encoding_weather, inplace=True)
df.head()

SEVERITYCODE  WEATHER ROADCOND                LIGHTCOND SPEEDING
0             2      3.0      Wet                 Daylight        N
1             1      6.0      Wet  Dark - Street Lights On        N
2             1      3.0      Dry                 Daylight        N
3             1      1.0      Dry                 Daylight        N
4             2      6.0      Wet                 Daylight        N

In [10]:
encoding_roadcond={'ROADCOND':{'Dry': 1, 'Sand/Mud/Dirt':2, 'Oil':3, 'Wet':4, 'Standing Water':5, 'Snow/Slush':6, 'Ice':7,'Other':8,'Unknown':9}}
df.replace(encoding_roadcond, inplace=True)
df.head()

SEVERITYCODE  WEATHER  ROADCOND                LIGHTCOND SPEEDING
0             2      3.0       4.0                 Daylight        N
1             1      6.0       4.0  Dark - Street Lights On        N
2             1      3.0       1.0                 Daylight        N
3             1      1.0       1.0                 Daylight        N
4             2      6.0       4.0                 Daylight        N

In [11]:
encoding_lightcond={'LIGHTCOND':{'Daylight': 1, 'Dawn':2, 'Dusk':3, 'Dark - Street Lights On':4, 'Dark - Street Lights Off':5,'Dark - No Street Lights':6,'Dark - Unknown Lighting':7,'Other':8,'Unknown':9}}
df.replace(encoding_lightcond, inplace=True)
df.head()

SEVERITYCODE  WEATHER  ROADCOND  LIGHTCOND SPEEDING
0             2      3.0       4.0        1.0        N
1             1      6.0       4.0        4.0        N
2             1      3.0       1.0        1.0        N
3             1      1.0       1.0        1.0        N
4             2      6.0       4.0        1.0        N

In [12]:
encoding_speed={'SPEEDING':{'Y':2,'N':1}}
df.replace(encoding_speed, inplace=True)
df.head()

SEVERITYCODE  WEATHER  ROADCOND  LIGHTCOND  SPEEDING
0             2      3.0       4.0        1.0         1
1             1      6.0       4.0        4.0         1
2             1      3.0       1.0        1.0         1
3             1      1.0       1.0        1.0         1
4             2      6.0       4.0        1.0         1

After that, lets balance out the dataset of SEVERITYCODE with 1 and 2 characters:

In [13]:
data_max= data[data.SEVERITYCODE==1]
data_min= data[data.SEVERITYCODE==2]

data_max_upsampled=resample(data_max,
                           replace=True,
                           n_samples=58188,
                           random_state=123)

balanced_df=pd.concat([data_max_upsampled, data_min])
balanced_df.SEVERITYCODE.value_counts()

2    58188
1    58188
Name: SEVERITYCODE, dtype: int64

After the Nan values or Null will be removed

In [14]:
df.dropna(axis='rows',inplace=True)

In [15]:
df.isnull().sum()

SEVERITYCODE    0
WEATHER         0
ROADCOND        0
LIGHTCOND       0
SPEEDING        0
dtype: int64

Now We can create an array with the new scale to study the data with dependent and independent variables, in this case lets use a matrix aproach:

In [16]:
x= np.asarray(df[['WEATHER', 'ROADCOND','LIGHTCOND','SPEEDING']])
x[0:5]

array([[3., 4., 1., 1.],
       [6., 4., 4., 1.],
       [3., 1., 1., 1.],
       [1., 1., 1., 1.],
       [6., 4., 1., 1.]])

In [17]:
y=df['SEVERITYCODE'].values
y[0:5]

array([2, 1, 1, 1, 2])

In [18]:
from sklearn import preprocessing
X = preprocessing.StandardScaler().fit(x).transform(x)
X[0:5]

array([[ 0.0104929 ,  0.65036952, -0.6551748 , -0.22752373],
       [ 1.05851211,  0.65036952,  0.66729801, -0.22752373],
       [ 0.0104929 , -0.61604977, -0.6551748 , -0.22752373],
       [-0.68818657, -0.61604977, -0.6551748 , -0.22752373],
       [ 1.05851211,  0.65036952, -0.6551748 , -0.22752373]])

As the Matrix has being placed, now lets split the dataset, 75% for trainning, 25% testing

In [19]:
from sklearn.model_selection import train_test_split
X_train, X_test,y_train,y_test=train_test_split(X,y,test_size=0.25, random_state=4)

print('Train set shape: ', X_train.shape,y_train.shape)
print('Test set shape: ', X_test.shape,y_test.shape)

Train set shape:  (142002, 4) (142002,)
Test set shape:  (47335, 4) (47335,)


### Models
This case, lets see different models behave with the dataset and can help to estimate the Severity of an accident. The models are:
<ul>
    <li>KNN</li>
    <li>Decision Tree</li>
    <li>SVM</li>
    <li>Logistic Regression</li>
</ul>


For the K Neighbor Model, lets try first with a number, this case 4; afterwards lets check it is an optimal point.

In [20]:
#KNN

from sklearn.neighbors import KNeighborsClassifier
k=4
neigh =KNeighborsClassifier(n_neighbors=k).fit(X_train, y_train)
neigh

Kyh= neigh.predict(X_test)
Kyh

print("Train set accuracy= ", metrics.accuracy_score(y_train, neigh.predict(X_train)))
print("Test set accuracy= ", metrics.accuracy_score(y_test, Kyh))

Train set accuracy=  0.6768989169166632
Test set accuracy=  0.6757156438153586


In [21]:
#Best KNN
Ks= 20

mean_acc = np.zeros((Ks-1))
std_acc=np.zeros((Ks-1))
ConfusionMx=[];

for n in range(1,Ks):
    neigh=KNeighborsClassifier(n_neighbors = n).fit(X_train, y_train)
    yh=neigh.predict(X_test)
    mean_acc[n-1]=metrics.accuracy_score(y_test,yh)
    std_acc[n-1]=np.std(yh==y_test)/np.sqrt(yh.shape[0])

mean_acc

array([0.65744164, 0.68433506, 0.62089363, 0.67571564, 0.62691455,
       0.66654695, 0.64715327, 0.65843456, 0.65085032, 0.68456744,
       0.6462026 , 0.66046266, 0.65332207, 0.66061054, 0.66052604,
       0.68549699, 0.6794972 , 0.685666  , 0.68040562])

And it was not! Usually is necessary to becarefull on not overfitting model and not forced the model to be a high number, though by the size of data set, and the time that may take to do more testing in the near future for more studies, in a debatle way lets take the number K=16 as good number. Nevertheless lets see how good this K number is by evaluating the model

In [22]:
#Decision Tree

from sklearn.tree import DecisionTreeClassifier
DT = DecisionTreeClassifier(criterion="entropy", max_depth =5)
DT.fit(X_train,y_train)

DTyh=DT.predict(X_test)

print("Train set accuracy= ", metrics.accuracy_score(y_train, DT.predict(X_train)))
print("Test set accuracy= ", metrics.accuracy_score(y_test, DTyh))

Train set accuracy=  0.6990957873832763
Test set accuracy=  0.6974331889722193


In [23]:
#SVM
from sklearn import svm

clf= svm.SVC(kernel='rbf')
clf.fit(X_train, y_train)

SVMyh=clf.predict(X_test)

print("Train set accuracy= ", metrics.accuracy_score(y_train, clf.predict(X_train)))
print("Test set accuracy= ", metrics.accuracy_score(y_test, SVMyh))

/opt/conda/envs/Python36/lib/python3.6/site-packages/sklearn/svm/base.py:196: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


Train set accuracy=  0.6990887452289405
Test set accuracy=  0.6974331889722193


In [26]:
#Logistic Regression

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix


LR= LogisticRegression(C=6,solver='liblinear').fit(X_train, y_train)

LRyh=LR.predict(X_test)
y_proba=LR.predict_proba(X_train)
print("Train set accuracy= ", metrics.accuracy_score(y_train, LR.predict(X_train)))
print("Test set accuracy= ", metrics.accuracy_score(y_test, LRyh))

Train set accuracy=  0.6990887452289405
Test set accuracy=  0.6974331889722193


### Results
According to the findings, We have got the following:
The best number for K Neighbor model would be K=4 due has ma better Jaccard and F1 score. Also, the resaults table the best models for future analysis with this dataset would be KNN and logistic regression if involving a binary answers for future questions that may occur. Also, needs to bring more attention on how significant would be for "car accidents" the "SPEEDING" variable. May help to kill the myth that high speed is bad?


In [27]:
#Model Evaluation

from sklearn.metrics import f1_score
from sklearn.metrics import jaccard_similarity_score
from sklearn.metrics import log_loss

print('KNN jaccard index: %.7f' % jaccard_similarity_score(y_test,Kyh))
print('KNN f1_score index: %.7f' % f1_score(y_test,Kyh, average='weighted'))

print('Decision Tree jaccard index: %.7f' % jaccard_similarity_score(y_test,DTyh))
print('Decision Tree f1_score index: %.7f' % f1_score(y_test,DTyh, average='weighted'))

print('SVM jaccard index: %.7f' % jaccard_similarity_score(y_test,SVMyh))
print('SVM f1_score index: %.7f' % f1_score(y_test,SVMyh, average='weighted'))

print('LogReg jaccard index: %.7f' % jaccard_similarity_score(y_test,LRyh))
print('LogReg f1_score index: %.7f' % f1_score(y_test,LRyh, average='weighted'))

print('LogReg logLoss: %.7f' %log_loss(y_train,y_proba))

KNN jaccard index: 0.6757156
KNN f1_score index: 0.6006030
Decision Tree jaccard index: 0.6974332
Decision Tree f1_score index: 0.5731160
SVM jaccard index: 0.6974332
SVM f1_score index: 0.5731160
LogReg jaccard index: 0.6974332
LogReg f1_score index: 0.5731160
LogReg logLoss: 0.5965423


/opt/conda/envs/Python36/lib/python3.6/site-packages/sklearn/metrics/classification.py:1143: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
/opt/conda/envs/Python36/lib/python3.6/site-packages/sklearn/metrics/classification.py:1143: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
/opt/conda/envs/Python36/lib/python3.6/site-packages/sklearn/metrics/classification.py:1143: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


| Algorithm          | Jaccard | F1-score | LogLoss |
| ------------------ | ------- | -------- | ------- |
| KNN                |0.6757   |0.6006    | NA      |
| Decision Tree      |0.6974   |0.5731    | NA      |
| SVM                |0.6974   |0.5731    | NA      |
| LogisticRegression |0.6974   |0.5731    | 0.5965  |

### Conclusion
After this study, what We can find, is that "to drive" is not a joke, is likely to have an accident, though no fatal, but hey, "who wants to give a shot for that, right?". So, it is important to consider the German model to provide driving licenses and at the same time to make sure that drivers give proper maintenance to their cars.

# Do not drink and Drive!